sktime/classification/deep_learning/base.py

```diff
@@ -123,3 +124,242 @@ def convert_y_to_keras(self, y):
+        """Magic method called during deserialization.
+
+        Parameters
+        ----------
+        state : dict, as returned from __getstate__(), used for correct deserialization
+
+        Returns
+        -------
+        -
+        """
+        from tensorflow.keras.optimizers import deserialize
+
+        self.__dict__ = state
+
+        if hasattr(self, "model_"):
+            self.__dict__["model_"] = self.model_
+            if hasattr(self, "model_.optimizer"):
+                self.__dict__["optimizer_"] = self.model_.optimizer
+
+        # if optimizer_ exists, set optimizer as optimizer_
+        if self.__dict__.get("optimizer_") is not None:
+            self.__dict__["optimizer"] = self.__dict__["optimizer_"]
+        # else model may not have been built, but an optimizer might be passed
+        else:
+            # Having 0 as value implies "optimizer" attribute was None
+            # as per __getstate__()
+            if self.__dict__.get("optimizer") == 0:
+                self.__dict__["optimizer"] = None
+            elif self.__dict__.get("optimizer") == -1:
+                # `optimizer` doesn't exist as a parameter alone, so delete it.
+                del self.__dict__["optimizer"]
+            else:
+                if isinstance(self.optimizer, dict):
+                    self.__dict__["optimizer"] = deserialize(self.optimizer)
+                else:
+                    # must have been a string already, no need to set
+                    pass
+
+        if hasattr(self, "history"):
+            self.__dict__["history"] = self.history
+
+    def save(self, path=None):
+        """Save serialized self to bytes-like object or to (.zip) file.
+
+        Behaviour:
+        if `path` is None, returns an in-memory serialized self
+        if `path` is a file, stores the zip with that name at the location.
+        The contents of the zip file are:
+        _metadata - contains class of self, i.e., type(self).
+        _obj - serialized self. This class uses the default serialization (pickle).
+        keras/ - model, optimizer and state stored inside this directory.
+        history - serialized history object.
+
+
+        Parameters
+        ----------
+        path : None or file location (str or Path)
+            if None, self is saved to an in-memory object
+            if file location, self is saved to that file location. For eg:
+                path="estimator" then a zip file `estimator.zip` will be made at cwd.
+                path="/home/stored/estimator" then a zip file `estimator.zip` will be
+                stored in `/home/stored/`.
+
+        Returns
+        -------
+        if `path` is None - in-memory serialized self
+        if `path` is file location - ZipFile with reference to the file
+        """
+        import pickle
+        import shutil
+        from pathlib import Path
+        from zipfile import ZipFile
+
+        if path is None:
+            import h5py
```

Comment:
we need to make sure we have proper isolation of soft dependencies.
This would simply break at use and test.

For use, I would put a check and an error message, i.e., "in-memory serialization requires `h5py`" or similar.

For the testing, we need to think about what to skip when.
The serialization test (save/load to memory) will fail without `h5py`. Since it is part of the standard test suite, this creates a problem for which I can't see a nice solution.

Reply:
(the general requirement being, the full test suite of sktime should run without failures with any set of soft dependencies installed)

Reply:
`tensorflow` is also a part of the soft dependencies, right? Or we do not consider it since it's also included in `dl` package?